File: apps/web/src/lib/onboarding.ts
export type OnboardingProfile = {
  employment_status: 'employee' | 'freelancer' | 'both' | 'retired' | 'student' | 'unemployed';
  marital_status: 'single' | 'married';
  has_children: boolean;
  num_children: number;
  rental_income: boolean;
  capital_gains: boolean;
  foreign_income: boolean;
  home_office_days: number;
  commute_km: number;
  commute_days: number;
  donations_eur: number;
  has_insurance_expenses: boolean;
  has_haushaltsnahe: boolean;
  has_handwerker: boolean;
  church_member: boolean;
  bundesland: string;
  tax_class: number;
};

export const EMPTY_PROFILE: OnboardingProfile = {
  employment_status: 'employee',
  marital_status: 'single',
  has_children: false,
  num_children: 0,
  rental_income: false,
  capital_gains: false,
  foreign_income: false,
  home_office_days: 0,
  commute_km: 0,
  commute_days: 230,
  donations_eur: 0,
  has_insurance_expenses: true,
  has_haushaltsnahe: false,
  has_handwerker: false,
  church_member: false,
  bundesland: 'Bayern',
  tax_class: 1,
};

export const BUNDESLAENDER = [
  'Baden-Wuerttemberg',
  'Bayern',
  'Berlin',
  'Brandenburg',
  'Bremen',
  'Hamburg',
  'Hessen',
  'Mecklenburg-Vorpommern',
  'Niedersachsen',
  'Nordrhein-Westfalen',
  'Rheinland-Pfalz',
  'Saarland',
  'Sachsen',
  'Sachsen-Anhalt',
  'Schleswig-Holstein',
  'Thueringen',
];

export function determineRequiredAnlagen(profile: OnboardingProfile): string[] {
  const required = ['Mantelbogen'];

  if (['employee', 'both', 'retired'].includes(profile.employment_status)) {
    required.push('Anlage N');
  }

  if (profile.employment_status === 'student') {
    // Students with Werkstudent/Minijob income may still need Anlage N
  }

  if (profile.employment_status === 'unemployed') {
    // ALG is tax-free but subject to Progressionsvorbehalt
  }

  if (profile.rental_income) {
    required.push('Anlage V');
  }

  if (profile.capital_gains) {
    required.push('Anlage KAP');
  }

  if (profile.has_children) {
    required.push('Anlage Kind');
  }

  if (profile.foreign_income) {
    required.push('Anlage AUS');
  }

  if (profile.has_insurance_expenses || ['employee', 'both', 'retired', 'student', 'unemployed'].includes(profile.employment_status)) {
    required.push('Anlage Vorsorge');
  }

  return required;
}

export function suggestQuickWins(profile: OnboardingProfile): string[] {
  const tips: string[] = [];

  if (profile.home_office_days > 0) {
    const capped = Math.min(profile.home_office_days, 210);
    tips.push(`Home office allowance: up to ${capped * 6} EUR (${capped} days x 6 EUR)`);
  }

  if (profile.commute_km > 0) {
    tips.push(`Commuting costs: Anlage N, Zeile 31 (${profile.commute_km} km)`);
  }

  if (profile.donations_eur > 0) {
    tips.push(`Donations of ${profile.donations_eur} EUR claimable as Sonderausgaben`);
  }

  if (profile.has_children) {
    tips.push('Childcare costs: 2/3 deductible, max 4,000 EUR per child');
  }

  if (profile.has_haushaltsnahe) {
    tips.push('Household services: 20% tax credit, max 4,000 EUR');
  }

  if (profile.has_handwerker) {
    tips.push('Craftsmen services: 20% tax credit, max 1,200 EUR');
  }

  tips.push('Bank account fees: flat 16 EUR deductible without receipts');

  return tips;
}

File: apps/api/app/domain/onboarding.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class OnboardingProfile:
    employment_status: str
    marital_status: str = "single"
    has_children: bool = False
    num_children: int = 0
    rental_income: bool = False
    capital_gains: bool = False
    foreign_income: bool = False
    home_office_days: int = 0
    commute_km: int = 0
    commute_days: int = 230
    donations_eur: float = 0.0
    has_insurance_expenses: bool = True
    has_haushaltsnahe: bool = False
    has_handwerker: bool = False
    church_member: bool = False
    bundesland: str = "Bayern"
    tax_class: int = 1


@dataclass(frozen=True)
class OnboardingResult:
    required_anlagen: list[str] = field(default_factory=list)
    suggested_next_steps: list[str] = field(default_factory=list)


def determine_required_anlagen(profile: OnboardingProfile) -> OnboardingResult:
    required_anlagen = ["Mantelbogen"]
    next_steps = ["Confirm personal data", "Review income sources"]

    if profile.employment_status in {"employee", "both", "retired"}:
        required_anlagen.append("Anlage N")
        next_steps.append("Prepare your Lohnsteuerbescheinigung")

    if profile.employment_status == "student":
        next_steps.append("Check if you had taxable income (Minijob, Werkstudent, internship)")
        next_steps.append("Students can carry forward losses (Verlustvortrag) for future use")

    if profile.employment_status == "unemployed":
        next_steps.append("Gather Arbeitslosengeld / ALG-I or ALG-II statements")
        next_steps.append("ALG is tax-free but subject to Progressionsvorbehalt")

    if profile.rental_income:
        required_anlagen.append("Anlage V")
        next_steps.append("Prepare rental expense documents")

    if profile.capital_gains:
        required_anlagen.append("Anlage KAP")
        next_steps.append("Collect broker statements")

    if profile.has_children:
        required_anlagen.append("Anlage Kind")
        next_steps.append("Gather child care cost receipts")

    if profile.foreign_income:
        required_anlagen.append("Anlage AUS")
        next_steps.append("Check foreign income disclosures")

    if profile.has_insurance_expenses or profile.employment_status in {"employee", "both", "retired", "student", "unemployed"}:
        required_anlagen.append("Anlage Vorsorge")
        next_steps.append("Collect insurance certificates")

    if profile.donations_eur > 0:
        next_steps.append("Add donations under Sonderausgaben")

    if profile.has_haushaltsnahe or profile.has_handwerker:
        next_steps.append("Enter household-related expenses (Haushaltsnahe)")

    if profile.home_office_days > 0:
        next_steps.append("Enter home office days for deduction")

    if profile.commute_km > 0:
        next_steps.append("Enter commute details for Entfernungspauschale")

    return OnboardingResult(required_anlagen=required_anlagen, suggested_next_steps=next_steps)
